core/client/fs/src/main/java/alluxio/client/file/FileSystemContext.java

```diff
@@ -271,95 +268,135 @@ public synchronized void close() throws IOException {
+      }
+      try {
+        getClientContext().loadConf(masterAddr, updateClusterConf, updatePathConf);
+      } catch (AlluxioStatusException e) {
+        // Failed to load configuration from meta master, maybe master is being restarted,
+        // or their is a temporary network problem, give up reinitialization. The heartbeat thread
+        // will try to reinitialize in the next heartbeat.
+        throw new UnavailableException(String.format("Failed to load configuration from meta master"
+            + " {} during reinitialization", masterAddr), e);
+      }
+      closeWithoutReinitializer();
+      initWithoutReinitializer(getClientContext(),
+          MasterInquireClient.Factory.create(getClusterConf()));
+      mReinitializer.onSuccess();
     }
   }
 
+  /**
+   * @return the unique ID of this context
+   */
+  public String getId() {
+    return mId;
+  }
+
+  /**
+   * @return the {@link MasterClientContext} backing this context
+   */
+  public MasterClientContext getMasterClientContext() {
+    return mMasterClientContext;
+  }
+
   /**
    * @return the {@link ClientContext} backing this {@link FileSystemContext}
    */
   public ClientContext getClientContext() {
-    return mClientContext;
+    return mMasterClientContext;
   }
 
   /**
    * @return the cluster level configuration backing this {@link FileSystemContext}
    */
   public AlluxioConfiguration getClusterConf() {
-    return mClientContext.getConf();
+    return getClientContext().getClusterConf();
   }
 
   /**
    * The path level configuration is a {@link SpecificPathConfiguration}.
    *
+   * If path level configuration has never been loaded from meta master yet, it will be loaded.
+   *
    * @param path the path to get the configuration for
    * @return the path level configuration for the specific path
    */
   public AlluxioConfiguration getPathConf(AlluxioURI path) {
-    return new SpecificPathConfiguration(mClientContext.getConf(), mClientContext.getPathConf(),
-        path);
+    return new SpecificPathConfiguration(getClientContext().getClusterConf(),
+        getClientContext().getPathConf(), path);
   }
 
   /**
    * @return the master address
    * @throws UnavailableException if the master address cannot be determined
    */
   public synchronized InetSocketAddress getMasterAddress() throws UnavailableException {
-    return mMasterInquireClient.getPrimaryRpcAddress();
+    return mMasterClientContext.getMasterInquireClient().getPrimaryRpcAddress();
   }
 
-  /**
-   * Acquires a file system master client from the file system master client pool.
-   *
-   * @return the acquired file system master client
-   */
-  public FileSystemMasterClient acquireMasterClient() {
-    try (LockResource r = new LockResource(mReinitLock.readLock())) {
+  private FileSystemMasterClient acquireMasterClient() {
+    try (LockResource r = acquireBlockReinitLockResource()) {
```

Comment:
is this redundant?

Reply:
before acquiring or releasing, we always acquire the lock, because otherwise, reinitialization might be happening concurrently, so the master client pool might have been closed concurrently.